### PR TITLE
compute: configure DD's idle_merge_effort

### PR DESCRIPTION
**Pull request by teskje**
This PR sets the `idle_merge_effort` configuration option of differential dataflow, which controls how much work DD spends on compacting arrangements. Anecdotically, the value 1000 provides a good trade-of between CPU time spent compacting more eagerly and memory saved by compacting earlier.

We might want to tune this value later, but it is clear that it should not be `None`. Setting the option to `None` means that arrangement compaction only happens in response to new updates arriving, so if a dataflow receives no updates for a while, its arrangements may sit around uncompacted, wasting precious memory.

### Motivation

  * This PR adds a known-desirable feature.

Advances #15974 by ensuring arrangements are compacted quickly during hydration.

### Tips for Reviewers

I thought about adding a CLI flag to control this setting, but decided not to because that would likely cause conflicts for Nikhil's storage-compute unification PR. We can always add it later, should people think it would be useful.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by teskje** [comment]
I added @philip-stoev because this has the potential to affect compute dataflow performance. The hope is that it doesn't, but I haven't checked yet. I'm planning to do so tomorrow using @vmarcos TPC-H benchmarks.

**Review by vmarcos — APPROVED** [review]
LGTM! Just let me know if you need any input regarding running TPC-H benchmark queries.

Some of the testing done for #16473 may be interesting, i.e., creating every TPC-H query as an indexed view and then trying to hydrate the system with an as large as possible scale factor for your machine. You can probably trigger similar behaviors by running queries concurrently without PK and FK indexes, i.e., forcing data to be read out of `persist`.

**Comment by teskje** [comment]
Unfortunately, both the feature benchmarks and TPC-H suggest that this change degrades performance.

## Feature benchmarks

The feature benchmarks crash at some point (timing out when trying to create a Kafka topic), but before that they output this:

```
NAME                      |    THIS     |    OTHER    |  Regression?  | 'THIS' is:
----------------------------------------------------------------------------------------------------
FastPathFilterNoIndex     |       0.317 |       0.317 |      no       | 0.1 pct   slower
FastPathFilterIndex       |       2.274 |       2.294 |      no       | 0.9 pct   faster
FastPathOrderByLimit      |       0.169 |       0.169 |      no       | 0.2 pct   faster
Insert                    |       3.510 |       2.741 |    !!YES!!    | 28.1 pct   slower
InsertBatch               |       7.296 |       7.314 |      no       | 0.2 pct   faster
InsertMultiRow            |       0.066 |       0.065 |      no       | 1.6 pct   slower
Update                    |       4.909 |       4.777 |      no       | 2.8 pct   slower
UpdateMultiNoIndex        |       5.971 |       5.102 |    !!YES!!    | 17.0 pct   slower
InsertAndSelect           |       4.419 |       3.364 |    !!YES!!    | 31.4 pct   slower
OrderBy                   |       9.234 |       8.788 |      no       | 5.1 pct   slower
CountDistinct             |       0.596 |       0.581 |      no       | 2.6 pct   slower
MinMax                    |       0.701 |       0.697 |      no       | 0.5 pct   slower
GroupBy                   |       3.199 |       3.117 |      no       | 2.6 pct   slower
CrossJoin                 |       1.701 |       1.712 |      no       | 0.7 pct   faster
Retraction                |       2.287 |       2.395 |      no       | 4.5 pct   faster
CreateIndex               |       1.090 |       1.061 |      no       | 2.7 pct   slower
DeltaJoin                 |       0.590 |       0.586 |      no       | 0.7 pct   slower
DifferentialJoin          |       0.528 |       0.476 |    !!YES!!    | 11.0 pct   slower
FullOuterJoin             |       4.734 |       4.908 |      no       | 3.6 pct   faster
FinishOrderByLimit        |       0.485 |       0.471 |      no       | 3.0 pct   slower
```

Some of the timings are significantly slower than on main, which can probably not be attributed to noise in the measurement.

## TCP-H

For TCP-H, I read @vmarcos comment to late and instead come up with the following:

    * Instantiate a TCP-H load generator source with tick interval 1s and scale factor 1.
    * Create PK and FK indexes on all input relations.
    * Create views with default indexes on all TCP-H queries.
    * Check CPU and memory usage.
    
I did all this on an xsmall instance (`default.r1` replica).

This is not a load test, but it includes a lot of arrangements, so I expected it to show the effects of the change in this PR well.

### Results

Setting `idle_merge_effort` does indeed reduce memory usage, but it increases CPU usage massively. This is clearly visible in the Grafana metrics:

* Without `idle_merge_effort`:
    Screenshot 2022-12-22 at 13 29 05
* With `idle_merge_effort`:
    Screenshot 2022-12-22 at 13 28 06

Memory usage decreases from 60 to 40%, while CPU usage increases from 6% to 100% (we only allocate half of the logical cores to timely, so the CPU graph is misleading). Looking at the flame graph, the bulk of CPU time is indeed spent in arrangement maintenance:

Screenshot 2022-12-22 at 13 55 54

The jump in CPU usage is way higher than I'd hoped. I didn't test different values for `idle_merge_effort`, it's possible that a smaller one might work out better. In any case, we should probably not merge this PR as-is.

**Comment by frankmcsherry** [comment]
Would it make sense to offer it as a replica creation argument?

**Comment by teskje** [comment]
> Would it make sense to offer it as a replica creation argument?

I created an issue to track this here: https://github.com/MaterializeInc/materialize/issues/16794

**Comment by teskje** [comment]
Unfortunately, an `idle_merge_effort` of 100 also doesn't make the feature benchmarks especially happy:

```
NAME                      |    THIS     |    OTHER    |  Regression?  | 'THIS' is:
----------------------------------------------------------------------------------------------------
FastPathFilterNoIndex     |       0.313 |       0.301 |      no       | 3.9 pct   slower
FastPathFilterIndex       |       2.250 |       2.242 |      no       | 0.4 pct   slower
FastPathOrderByLimit      |       0.170 |       0.169 |      no       | 0.3 pct   slower
Insert                    |       3.742 |       2.808 |    !!YES!!    | 33.3 pct   slower
InsertBatch               |       7.311 |       7.389 |      no       | 1.1 pct   faster
InsertMultiRow            |       0.065 |       0.066 |      no       | 2.5 pct   faster
Update                    |       4.928 |       4.700 |      no       | 4.8 pct   slower
UpdateMultiNoIndex        |       5.997 |       5.091 |    !!YES!!    | 17.8 pct   slower
InsertAndSelect           |       4.335 |       3.391 |    !!YES!!    | 27.8 pct   slower
OrderBy                   |       8.407 |       8.776 |      no       | 4.2 pct   faster
CountDistinct             |       0.649 |       0.609 |      no       | 6.5 pct   slower
MinMax                    |       0.705 |       0.703 |      no       | 0.4 pct   slower
GroupBy                   |       3.191 |       3.173 |      no       | 0.6 pct   slower
CrossJoin                 |       1.857 |       1.825 |      no       | 1.7 pct   slower
Retraction                |       2.288 |       2.386 |      no       | 4.1 pct   faster
CreateIndex               |       1.088 |       1.072 |      no       | 1.5 pct   slower
DeltaJoin                 |       0.604 |       0.598 |      no       | 1.0 pct   slower
DifferentialJoin          |       0.520 |       0.491 |      no       | 6.0 pct   slower
FullOuterJoin             |       5.207 |       4.901 |      no       | 6.3 pct   slower
FinishOrderByLimit        |       0.468 |       0.482 |      no       | 2.8 pct   faster
```

**Comment by frankmcsherry** [comment]
Hrm. I wonder if the benchmarks are just a bit weird / pessimal. The "insert" benchmarks seem likely to trickle in changes, which will cause a relatively larger amount of CPU to be used merging. Mysterious! I'm surprised that inserting takes more time, but I suspect we are measuring more than just insertions (also reading it back, maybe?).

**Comment by teskje** [comment]
Here is `idle_merge_effort = 10`, for good measure:
```
NAME                      |    THIS     |    OTHER    |  Regression?  | 'THIS' is:
----------------------------------------------------------------------------------------------------
FastPathFilterNoIndex     |       0.316 |       0.306 |      no       | 3.2 pct   slower
FastPathFilterIndex       |       2.268 |       2.278 |      no       | 0.4 pct   faster
FastPathOrderByLimit      |       0.169 |       0.170 |      no       | 0.4 pct   faster
Insert                    |       3.702 |       2.789 |    !!YES!!    | 32.7 pct   slower
InsertBatch               |       7.305 |       7.365 |      no       | 0.8 pct   faster
InsertMultiRow            |       0.067 |       0.067 |      no       | 0.9 pct   slower
Update                    |       5.393 |       4.719 |    !!YES!!    | 14.3 pct   slower
UpdateMultiNoIndex        |       5.977 |       5.085 |    !!YES!!    | 17.5 pct   slower
InsertAndSelect           |       4.192 |       3.394 |    !!YES!!    | 23.5 pct   slower
OrderBy                   |       8.768 |       8.537 |      no       | 2.7 pct   slower
CountDistinct             |       0.611 |       0.655 |      no       | 6.7 pct   faster
MinMax                    |       0.687 |       0.720 |      no       | 4.6 pct   faster
GroupBy                   |       3.240 |       3.098 |      no       | 4.6 pct   slower
CrossJoin                 |       1.691 |       1.785 |      no       | 5.2 pct   faster
Retraction                |       2.593 |       2.250 |    !!YES!!    | 15.2 pct   slower
CreateIndex               |       1.076 |       1.027 |      no       | 4.8 pct   slower
DeltaJoin                 |       0.597 |       0.588 |      no       | 1.5 pct   slower
DifferentialJoin          |       0.517 |       0.518 |      no       | 0.1 pct   faster
FullOuterJoin             |       4.945 |       5.239 |      no       | 5.6 pct   faster
FinishOrderByLimit        |       0.476 |       0.460 |      no       | 3.6 pct   slower
```

Strangely, it seems to be worse than 100. Maybe I'm doing something wrong when running these benchmarks, I should probably check what exactly they are doing.

**Comment by vmarcos** [comment]
It might be more reliable to give the TPC-H scenario another run? It looks like the differences in the feature benchmark are on the one hand measurable, but on the other hand also not that frighteningly large.

**Comment by philip-stoev** [comment]
@teskje if this is observable with the TPC-H , you may wish to consider using it for tuning going forward.

Re the feature benchmark, when Platform was introduced the numbers started exhibiting massive variability. In order to counter that, the feature benchmark in CI is set to retry to regression = Yes scenarios up to 5 times to weed out any false positives. However, if the feature benchmark crashes in the middle of things than this retrying will not happen. I am now going to fire up a run in the CI and see if any more authoritative numbers could be produced.

**Comment by philip-stoev** [comment]
I ran the Feature Benchmark using CI and the retries mentioned above - https://buildkite.com/materialize/nightlies/builds/1559#01855304-700e-4771-84f7-5168968f3382

The results indicate no performance regression is detectable with the Feature Benchmark that is specific to this branch:
- most regressions reported were false positives that went away on retr
- the ParallelDataflows regression is also present in `main`

Apologies for the confusion.

**Comment by teskje** [comment]
Thanks Philip!

As suggested, I did some more TPC-H benchmarking. This time I created a one-off TPC-H load generator and measured the time it took to run the queries Q01 to Q22 in sequence. I did this with no indexes and with the full set of pk/fk indexes.

Here are the results (times are minimum values of two runs each):

* Without indexes:

  |    | idle_merge_effort = None | idle_merge_effort = 1000 |
  |---|---|---|
  |  total query time | 266 s | 255 s | 
  |  max memory usage | 2.16 GB | 2.56 GB |
  |  max CPU usage | 80% | 83% |

* With indexes:

  |    | idle_merge_effort = None | idle_merge_effort = 1000 |
  |---|---|---|
  |  total query time | 42 s | 37 s | 
  |  max memory usage | 7.17 GB | 7.78 GB |
  |  max CPU usage | 80% | 100% |

Queries under idle_merge_effort are faster than queries without idle_merge_effort. The difference is not huge, but it is consistent from what I can tell, also between individual TPC-H queries.

As expected, setting idle_merge_effort leads to higher CPU usage.

What's unexpected is that setting idle_merge_effort also leads to higher memory usage. We would expect to see the opposite. But then again the differences are not huge and I simply eyeballed these metrics from Grafana, so it is entirely possible that this is just noise. My previous TPC-H benchmark showed that there can be a significant reduction in memory usage from enabling idle_merge_effort.

Based on these results, and my previous smooth-rehydration experiments, it does seem like setting idle_merge_effort does have a benefit. So I think it's fine to merge this PR, but I'm also happy to do further tests if anyone would like to suggest some!

**Comment by vmarcos** [comment]
I think that the data we have so far confirm that setting `idle_merge_effort` at all, and to 1000 in particular, provides a stability benefit (in the form of lower memory consumption) during hydration and a (small) performance benefit during query processing compared with not setting the parameter. So if the question we were aiming at was "Should we set `idle_merge_effort`, for example to a 1000 by default?" then I think the answer is probably yes, and we should merge the PR as is.

However, if we were trying to answer the questions "What is a good default value for idle_merge_effort?" or "How do I set idle_merge_effort?" then I think we'd need more experimentation. Perhaps answering these questions should be the subject of a follow-up issue/epic?

I can see at least the following characteristics that we might want to vary using single-factor style for experiments:

1. The value of `idle_merge_effort`, e.g., 0, 10, 100, 1000.
2. The tick interval (thus the update rate) in the TPC-H load generator, e.g., none, 0.001s, 0.01s, 0.1s, 1s. Without updates, it may be that we do not have a lot to merge to start with!
3. The level of concurrency in the workload, e.g., you saw a big benefit in memory usage with concurrent hydration of all queries and a degradation in memory usage in sequential execution of the TPC-H queries.
4. The use of persist to read in the data vs. employing a fully indexed design.

**Comment by teskje** [comment]
Thanks @vmarcos! I agree that we should do more testing to see if there is a better default. I created a follow-up issue: https://github.com/MaterializeInc/materialize/issues/16906.

**Pull request merged**
